scylla/src/transport/locator/tablets.rs

```diff
@@ -0,0 +1,630 @@
+    }
+
+    #[test]
+    fn test_adding_tablet_overlapping_one() {
+        let mut tablets = Default::default();
+        insert_ranges(&mut tablets, &[(-2000000, -1800000)]);
+
+        // Replacing a tablet, overlaps right part of the old one
+        insert_ranges(&mut tablets, &[(-1900000, -1700000)]);
+        verify_ranges(&tablets, &[(-1900000, -1700000)]);
+
+        // Replacing a tablet, overlaps left part of the old one
+        insert_ranges(&mut tablets, &[(-2000000, -1800000)]);
+        verify_ranges(&tablets, &[(-2000000, -1800000)]);
+    }
+
+    #[test]
+    fn test_adding_tablet_fill_hole() {
+        let mut tablets = Default::default();
+
+        // Fill a hole between two tablets
+        insert_ranges(
+            &mut tablets,
+            &[
+                (-2000000, -1800001),
+                (-1600000, -1400000), // Create a hole
+                (-1800000, -1600001), // Fully fill this hole
+            ],
+        );
+        verify_ranges(
+            &tablets,
+            &[
+                (-2000000, -1800001),
+                (-1800000, -1600001),
+                (-1600000, -1400000),
+            ],
+        );
+    }
+
+    #[test]
+    fn test_adding_tablet_neighbours_not_removed() {
+        let mut tablets = Default::default();
+        insert_ranges(
+            &mut tablets,
+            &[
+                (-2000000, -1800001),
+                (-1800000, -1600001),
+                (-1600000, -1400000),
+            ],
+        );
+
+        // Make sure neighbours are not removed when fully replacing tablet in the middle
+        insert_ranges(&mut tablets, &[(-1800000, -1600001)]);
+        verify_ranges(
+            &tablets,
+            &[
+                (-2000000, -1800001),
+                (-1800000, -1600001),
+                (-1600000, -1400000),
+            ],
+        );
+
+        // Make sure neighbours are not removed when new tablet is smaller than old one
+        insert_ranges(&mut tablets, &[(-1750000, -1650000)]);
+        verify_ranges(
+            &tablets,
+            &[
+                (-2000000, -1800001),
+                (-1750000, -1650000),
+                (-1600000, -1400000),
+            ],
+        );
+    }
+
+    #[test]
+    fn replace_multiple_tablets_middle() {
+        let mut tablets = Default::default();
+        insert_ranges(
+            &mut tablets,
+            &[
+                (-2000000, -1800001),
+                (-1800000, -1600001),
+                (-1600000, -1400001),
+                (-1400000, -1200001),
+                (-1200000, -1000000),
+            ],
+        );
+
+        // Replacing 3 middle tablets
+        insert_ranges(&mut tablets, &[(-1750000, -1250000)]);
+        verify_ranges(
+            &tablets,
+            &[
+                (-2000000, -1800001),
+                (-1750000, -1250000),
+                (-1200000, -1000000),
+            ],
+        );
+    }
+}
```

Comment:
I added the following comment to `TableTablets` struct:
```
/// Container for tablets of a single table.
/// 
/// It can be viewed as a set of non-overlapping Tablet objects.
/// It has 2 basic operations:
/// 1. Find a tablet for given Token
/// 2. Add a new tablet.
/// 
/// Adding new Tablet will first remove all tablets that overlap with the new tablet.
```

is this explanation enough?